File: frontend/src/utils/pageCounter.js
/**
 * Frontend Page Counter Utility
 * 
 * Client-side estimates for report page counts.
 * Aligns with backend page_counter.py for consistency.
 * 
 * Note: Backend provides accurate counts via JSON metadata.
 * This utility is for UI estimation and fallback purposes.
 */

// A4 dimensions in pixels (at 96dpi)
const A4_HEIGHT_PX = 1122; // 297mm
const A4_WIDTH_PX = 794;   // 210mm
const MARGIN_TOP_PX = 20;
const MARGIN_BOTTOM_PX = 20;
const MARGIN_LEFT_PX = 20;
const MARGIN_RIGHT_PX = 20;

const USABLE_HEIGHT = A4_HEIGHT_PX - MARGIN_TOP_PX - MARGIN_BOTTOM_PX; // 1082px
const USABLE_WIDTH = A4_WIDTH_PX - MARGIN_LEFT_PX - MARGIN_RIGHT_PX;   // 754px

const HEADER_HEIGHT = 150;  // Logo + title in pixels
const FOOTER_HEIGHT = 50;   // Page number, etc.
const CONTENT_HEIGHT = USABLE_HEIGHT - HEADER_HEIGHT - FOOTER_HEIGHT; // ~882px

/**
 * Calculate pages for Ledger Report
 * ~25-28 rows per page (each row ~8mm / ~30px)
 */
export function calculateLedgerPages(recordCount) {
  if (recordCount === 0) return 1;
  
  const rowsPerPage = 27;
  const pages = Math.ceil(recordCount / rowsPerPage);
  
  return Math.max(1, pages);
}

/**
 * Calculate pages for Group Total Report
 * ~20-22 groups per page (each group row ~10mm / ~38px)
 */
export function calculateGroupTotalPages(groupCount) {
  if (groupCount === 0) return 1;
  
  const groupsPerPage = 21;
  const pages = Math.ceil(groupCount / groupsPerPage);
  
  return Math.max(1, pages);
}

/**
 * Calculate pages for Daily Sales Report
 * ~32-35 rows per page (compact layout, ~7mm / ~26px per row)
 */
export function calculateDailySalesPages(recordCount) {
  if (recordCount === 0) return 1;
  
  const rowsPerPage = 33;
  const pages = Math.ceil(recordCount / rowsPerPage);
  
  return Math.max(1, pages);
}

/**
 * Calculate pages for Group Patti Report (most complex)
 * 
 * Per group with 15 entries: ~450px (fits ~2 groups per page)
 * Per group with 30 entries: ~820px (fits ~1 group per page)
 * Per group with 50+ entries: requires 2+ pages
 * 
 * @param {number} totalRecords - Total transactions across all groups
 * @param {number} groupCount - Number of groups
 * @param {number} avgRecordsPerGroup - Average records per group (optional)
 * @returns {number} Estimated page count
 */
export function calculateGroupPattiPages(totalRecords, groupCount, avgRecordsPerGroup = null) {
  if (totalRecords === 0) return 1;
  if (groupCount === 0) return 1;
  
  // Calculate average if not provided
  const avgRecords = avgRecordsPerGroup || (totalRecords / groupCount);
  
  // Each group occupies:
  // - Header: 30px
  // - Footer: 30px
  // - Gap: 20px
  // - Detail rows: 7mm (~26px) each
  
  const groupContentHeight = 80; // header + footer + gap
  const perRecordHeight = 26;
  
  const avgGroupHeight = groupContentHeight + (avgRecords * perRecordHeight);
  
  if (avgGroupHeight < CONTENT_HEIGHT) {
    const groupsPerPage = Math.floor(CONTENT_HEIGHT / avgGroupHeight);
    const pages = Math.ceil(groupCount / Math.max(1, groupsPerPage));
    return Math.max(1, pages);
  } else {
    // Large groups: each group may need multiple pages
    const pagesPerGroup = Math.ceil(avgGroupHeight / CONTENT_HEIGHT);
    const pages = groupCount * pagesPerGroup;
    return Math.max(1, pages);
  }
}

/**
 * Unified function to estimate page count
 * Aligns with backend estimate_pdf_page_count()
 * 
 * @param {string} reportType - 'ledger', 'group_total', 'daily_sales', or 'group_patti'
 * @param {number} recordCount - Number of detail records
 * @param {number} groupCount - Number of groups (for group_patti)
 * @param {number} avgRecordsPerGroup - Average records per group (for group_patti)
 * @returns {number} Estimated page count
 */
export function estimatePdfPageCount(
  reportType,
  recordCount = 0,
  groupCount = 0,
  avgRecordsPerGroup = null
) {
  switch (reportType) {
    case 'ledger':
      return calculateLedgerPages(recordCount);
    case 'group_total':
      return calculateGroupTotalPages(groupCount);
    case 'daily_sales':
      return calculateDailySalesPages(recordCount);
    case 'group_patti':
      return calculateGroupPattiPages(recordCount, groupCount, avgRecordsPerGroup);
    default:
      return 1;
  }
}

/**
 * Estimate page count from HTML content
 * Useful for fallback when metadata is unavailable
 * 
 * @param {string} htmlContent - HTML string to analyze
 * @returns {number} Estimated page count based on content height
 */
export function estimatePagesFromHtml(htmlContent) {
  try {
    // Create a temporary DOM element
    const parser = new DOMParser();
    const doc = parser.parseFromString(htmlContent, 'text/html');
    
    // Get total content height
    const pages = doc.querySelectorAll('.page');
    if (pages.length > 0) {
      return pages.length;
    }
    
    // Fallback: estimate from total body height
    const body = doc.body;
    if (body) {
      const heights = Array.from(doc.querySelectorAll('*')).map(el => {
        const style = window.getComputedStyle(el);
        const height = parseFloat(style.height);
        return isNaN(height) ? 0 : height;
      });
      
      const totalHeight = Math.max(...heights);
      return Math.ceil(totalHeight / A4_HEIGHT_PX) || 1;
    }
    
    return 1;
  } catch (error) {
    console.error('Error estimating pages from HTML:', error);
    return 1; // Fallback to 1 page
  }
}

/**
 * Format page count for display
 * 
 * @param {number} current - Current page number
 * @param {number} total - Total pages
 * @returns {string} Formatted string like "Page 3 of 10"
 */
export function formatPageCount(current, total) {
  return `Page ${Math.max(1, current)} of ${Math.max(1, total)}`;
}

/**
 * Check if page count is acceptable for browser preview
 * Large documents may cause performance issues
 * 
 * @param {number} pageCount - Number of pages
 * @param {number} warningThreshold - Warn if exceeds this (default: 50)
 * @param {number} errorThreshold - Error if exceeds this (default: 100)
 * @returns {Object} { status, message } where status is 'ok', 'warning', or 'error'
 */
export function validatePageCount(pageCount, warningThreshold = 50, errorThreshold = 100) {
  if (pageCount >= errorThreshold) {
    return {
      status: 'error',
      message: `Report has ${pageCount} pages. Browser preview may be slow. Consider printing to PDF directly.`
    };
  }
  
  if (pageCount >= warningThreshold) {
    return {
      status: 'warning',
      message: `Report has ${pageCount} pages. Browser preview may be slower. Consider opening in new tab.`
    };
  }
  
  return {
    status: 'ok',
    message: `Report ready (${pageCount} pages)`
  };
}

export default {
  calculateLedgerPages,
  calculateGroupTotalPages,
  calculateDailySalesPages,
  calculateGroupPattiPages,
  estimatePdfPageCount,
  estimatePagesFromHtml,
  formatPageCount,
  validatePageCount
};

File: backend/app/utils/page_counter.py
"""
Page counter utility for calculating expected page counts for various reports.
Uses A4 page dimensions with standard margins to estimate record-to-page ratio.
"""
import math


# A4 Page Dimensions (in mm)
A4_WIDTH = 210
A4_HEIGHT = 297
MARGIN_TOP = 5
MARGIN_BOTTOM = 5
MARGIN_LEFT = 5
MARGIN_RIGHT = 5

# Usable area
USABLE_WIDTH = A4_WIDTH - MARGIN_LEFT - MARGIN_RIGHT  # 200mm
USABLE_HEIGHT = A4_HEIGHT - MARGIN_TOP - MARGIN_BOTTOM  # 287mm

# Header/Footer space (logo, title, etc.) - approximate in mm
HEADER_HEIGHT = 40  # Logo + title
FOOTER_HEIGHT = 15  # Page number, etc.

# Available space for content
CONTENT_HEIGHT = USABLE_HEIGHT - HEADER_HEIGHT - FOOTER_HEIGHT  # ~232mm


def calculate_ledger_pages(record_count: int) -> int:
    """
    Calculate pages for Ledger Report (Silk Ledger Entries).
    
    Layout: Vertical table with columns for date, qty, kg, rate, amount
    Typical row height: ~8mm (with padding)
    Header row: ~10mm
    
    Per page capacity: ~25-28 rows
    """
    if record_count == 0:
        return 1
    
    rows_per_page = 27
    total_rows = record_count + 1  # +1 for header
    pages = math.ceil(record_count / rows_per_page)
    
    return max(1, pages)


def calculate_group_total_pages(group_count: int) -> int:
    """
    Calculate pages for Group Total Report.
    
    Layout: Summary table with groups and totals
    Typical row height: ~10mm (with padding)
    Header row: ~12mm
    
    Per page capacity: ~20-22 groups
    """
    if group_count == 0:
        return 1
    
    groups_per_page = 21
    pages = math.ceil(group_count / groups_per_page)
    
    return max(1, pages)


def calculate_daily_sales_pages(record_count: int) -> int:
    """
    Calculate pages for Daily Sales Report.
    
    Layout: Daily sales table with columns for date, party, item, qty, rate, total
    Typical row height: ~7mm (compact)
    Header row: ~9mm
    
    Per page capacity: ~32-35 rows
    """
    if record_count == 0:
        return 1
    
    rows_per_page = 33
    pages = math.ceil(record_count / rows_per_page)
    
    return max(1, pages)


def calculate_group_patti_pages(
    total_records: int,
    group_count: int,
    avg_records_per_group: float = None
) -> int:
    """
    Calculate pages for Group Patti Report (most complex).
    
    Layout: Multi-section report with:
    - Group header (8mm)
    - Detail rows (7mm each)
    - Group footer/summary (8mm)
    - Gap between groups (5mm)
    
    Per group with 15 entries: ~130mm (fits ~2 groups per page)
    Per group with 30 entries: ~235mm (fits ~1 group per page)
    Per group with 50+ entries: requires 2+ pages
    
    Args:
        total_records: Total transactions across all groups
        group_count: Number of groups
        avg_records_per_group: Average records per group (optional, auto-calculated)
    
    Returns:
        Estimated page count
    """
    if total_records == 0:
        return 1
    
    if group_count == 0:
        return 1
    
    # Calculate average if not provided
    if avg_records_per_group is None:
        avg_records_per_group = total_records / group_count
    
    # Each group occupies: header (8) + footer (8) + gap (5) + detail rows
    # Detail row height: 7mm
    group_content_height = 21  # header + footer + gap
    per_record_height = 7
    
    # Available space accounting for first page header
    first_page_available = CONTENT_HEIGHT
    subsequent_page_available = CONTENT_HEIGHT
    
    # Simple calculation: estimate groups per page
    # First group typically fits with header; subsequent groups depend on content
    avg_group_height = group_content_height + (avg_records_per_group * per_record_height)
    
    if avg_group_height < first_page_available:
        groups_per_page = int(first_page_available / avg_group_height)
        pages = math.ceil(group_count / max(1, groups_per_page))
    else:
        # Large groups: each group may need multiple pages
        pages_per_group = math.ceil(avg_group_height / subsequent_page_available)
        pages = group_count * pages_per_group
    
    return max(1, pages)


def estimate_pdf_page_count(
    report_type: str,
    record_count: int = 0,
    group_count: int = 0,
    avg_records_per_group: float = None
) -> int:
    """
    Unified function to estimate page count for any report type.
    
    Args:
        report_type: 'ledger', 'group_total', 'daily_sales', or 'group_patti'
        record_count: Number of detail records
        group_count: Number of groups (for group_patti)
        avg_records_per_group: Average records per group (for group_patti)
    
    Returns:
        Estimated page count
    """
    if report_type == "ledger":
        return calculate_ledger_pages(record_count)
    elif report_type == "group_total":
        return calculate_group_total_pages(group_count)
    elif report_type == "daily_sales":
        return calculate_daily_sales_pages(record_count)
    elif report_type == "group_patti":
        return calculate_group_patti_pages(record_count, group_count, avg_records_per_group)
    else:
        return 1
